### PR TITLE
Lpc55S36 flex pwm enablement

### PR DESCRIPTION
**Pull request by EmilioCBen**
The Enablement of FlexPWM for the LPC55S36 part, broken up into separate commits organized via folders

### CODE REVIEW

File: boards/arm/lpcxpresso55s36/lpcxpresso55s36.yaml [hunk 1]
```diff
@@ -16,3 +16,4 @@ toolchain:
   - xtools
 supported:
   - gpio
+  - flexpwm
```
Comment (galak) [inline boards/arm/lpcxpresso55s36/lpcxpresso55s36.yaml:19]: remove this.  If you want to add something, add `pwm`

File: drivers/pwm/pwm_mcux.c [hunk 2]
```diff
@@ -135,23 +141,42 @@ static int pwm_mcux_init(const struct device *dev)
...
 		return err;
 	}
-
+#endif
 	PWM_GetDefaultConfig(&pwm_config);
 	pwm_config.prescale = config->prescale;
 	pwm_config.reloadLogic = kPWM_ReloadPwmFullCycle;
 	pwm_config.clockSource = kPWM_BusClock;
 
+	/* @NOTE(Emilio): Set the Submodule Clocks for FlexPWM */
```
Comment (galak) [inline drivers/pwm/pwm_mcux.c:162]: Fix comment.

File: drivers/pwm/pwm_mcux.c [hunk 3]
```diff
@@ -135,23 +141,42 @@ static int pwm_mcux_init(const struct device *dev)
...
+#if CONFIG_BOARD_LPCXPRESSO55S36
+	SYSCON->PWM0SUBCTL |=
+	  (SYSCON_PWM0SUBCTL_CLK0_EN_MASK | SYSCON_PWM0SUBCTL_CLK1_EN_MASK |
+	   SYSCON_PWM0SUBCTL_CLK2_EN_MASK);
+	SYSCON->PWM1SUBCTL |=
+	  (SYSCON_PWM1SUBCTL_CLK0_EN_MASK | SYSCON_PWM1SUBCTL_CLK1_EN_MASK |
+	   SYSCON_PWM1SUBCTL_CLK2_EN_MASK);
+#endif
+	/* @NOTE(Emilio): Since the Idle function at the system's level of */
```
Comment (galak) [inline drivers/pwm/pwm_mcux.c:171]: Fix comment.

File: dts/arm/nxp/nxp_lpc55S3x_common.dtsi [hunk 4]
```diff
@@ -225,6 +225,109 @@
...
 
+	flexpwm0: flexpwm@400C3000 {
+		compatible = "nxp,flexpwm";
+		reg = <0x400C3000 0x1000>;
+		interrupt-names = "INPUT-CAPTURE", "FAULT", "RELOAD-ERROR";
+		interrupts = <80 0>, <81 0>, <82 0>;
+		flexpwm0_pwm0: pwm0 {
+			compatible = "nxp,imx-pwm";
+			index = <0>;
+			label = "FLEXPWM0_PWM0";
```
Comment (galak) [inline dts/arm/nxp/nxp_lpc55S3x_common.dtsi:236]: remove label

File: soc/arm/nxp_lpc/lpc55xxx/Kconfig.defconfig.lpc55S36 [hunk 5]
```diff
@@ -8,4 +8,15 @@ if SOC_LPC55S36
 config SOC
 	default "lpc55S36"
 
+config SOC_FLASH_MCUX
```
Comment (galak) [inline soc/arm/nxp_lpc/lpc55xxx/Kconfig.defconfig.lpc55S36:11]: remove this

File: soc/arm/nxp_lpc/lpc55xxx/Kconfig.defconfig.lpc55S36 [hunk 6]
```diff
@@ -8,4 +8,15 @@ if SOC_LPC55S36
...
+	default y
+	depends on FLASH
+
+config HAS_MCUX
+	default y
+
+config PWM_MCUX
```
Comment (galak) [inline soc/arm/nxp_lpc/lpc55xxx/Kconfig.defconfig.lpc55S36:18]: remove this, shouldn't be needed

File: soc/arm/nxp_lpc/lpc55xxx/Kconfig.defconfig.lpc55S36 [hunk 7]
```diff
@@ -8,4 +8,15 @@ if SOC_LPC55S36
 config SOC
 	default "lpc55S36"
 
+config SOC_FLASH_MCUX
+	default y
+	depends on FLASH
+
+config HAS_MCUX
```
Comment (galak) [inline soc/arm/nxp_lpc/lpc55xxx/Kconfig.defconfig.lpc55S36:15]: probably already done as `select` as well, so remove this.

**Review by galak — CHANGES_REQUESTED** [review]
fix dts/Kconfig

File: drivers/pwm/Kconfig.mcux [hunk 8]
```diff
@@ -2,10 +2,18 @@
 
 # Copyright (c) 2019 Linaro Limited.
 # SPDX-License-Identifier: Apache-2.0
+# Copyright (c) 2022, NXP
```
Comment (danieldegrasse) [inline drivers/pwm/Kconfig.mcux:5]: Copyright should be formatted as `Copyright 2022 NXP`

File: soc/arm/nxp_lpc/lpc55xxx/soc.h [hunk 9]
```diff
@@ -32,6 +32,8 @@
 #define IOCON_PIO_FUNC7 0x07u         /*!<@brief Selects pin function 7 */
 #define IOCON_PIO_FUNC9 0x09u         /*!<@brief Selects pin function 9 */
 #define IOCON_PIO_FUNC10 0x0Au        /*!<@brief Selects pin function 10 */
+#define IOCON_PIO_FUNC11 0x0Bu        /*!<@brief Selects pin function 11 */
```
Comment (danieldegrasse) [inline soc/arm/nxp_lpc/lpc55xxx/soc.h:35]: This change should either be in a separate commit, or the commit message needs to be updated to reflect that you are defining the missing PIO_FUNC11 definition.

File: drivers/pwm/Kconfig.mcux [hunk 10]
```diff
@@ -2,10 +2,18 @@
...
 config PWM_MCUX
 	bool "MCUX PWM driver"
 	default y
 	depends on DT_HAS_NXP_IMX_PWM_ENABLED
 	help
 	  Enable mcux pwm driver.
+
+config PWM_WAITEN_FLAG
+	bool "PWM Wait Enable Flag"
+	depends on HAS_MCUX_PWM
```
Comment (danieldegrasse) [inline drivers/pwm/Kconfig.mcux:16]: Rather than depending on `HAS_MCUX_PWM`, I would place this Kconfig within an `if PWM_MCU` block, to make it (and other future MCUX PWM specific Kconfigs) dependent on `PWM_MCUX`

File: boards/arm/lpcxpresso55s36/lpcxpresso55s36-pinctrl.dtsi [hunk 11]
```diff
@@ -42,10 +42,94 @@
 	/* Configures pin routing and optionally pin electrical features. */
 	pinmux_sctimer_default: pinmux_sctimer_default {
 		group0 {
-			pinmux = <SCT0_OUT0_PIO1_4>,
-				<SCT0_OUT2_PIO0_15>;
+		  pinmux = <FC2_TXD_SCL_MISO_WS_PIO1_25>;
```
Comment (danieldegrasse) [inline boards/arm/lpcxpresso55s36/lpcxpresso55s36-pinctrl.dtsi:45]: Why is the pinmux setting for sctimer changed here? Was the previous definition incorrect?

File: drivers/pwm/pwm_mcux.c [hunk 12]
```diff
@@ -135,23 +141,42 @@ static int pwm_mcux_init(const struct device *dev)
...
 	}
-
+#endif
 	PWM_GetDefaultConfig(&pwm_config);
 	pwm_config.prescale = config->prescale;
 	pwm_config.reloadLogic = kPWM_ReloadPwmFullCycle;
 	pwm_config.clockSource = kPWM_BusClock;
 
+	/* Set the Submodule Clocks for FlexPWM */
+#if CONFIG_BOARD_LPCXPRESSO55S36
```
Comment (galak) [inline drivers/pwm/pwm_mcux.c:163]: We shouldn't have board specific code in a driver

**Review by galak — DISMISSED** [review]
Need to figure out where there board specific code should go, but it shouldn't be in the driver.

File: drivers/pwm/pwm_mcux.c (hunk 12 above)
Comment (galak) [reply drivers/pwm/pwm_mcux.c:163]: Probably better placed in `soc/arm/nxp_lpc/lpc55xxx/soc.c` in `clock_init`

Comment (EmilioCBen) [reply drivers/pwm/pwm_mcux.c:163]: Ahh I see, thank you for the suggestion, will be fixed right away!

File: boards/arm/lpcxpresso55s36/lpcxpresso55s36.dts [hunk 13]
```diff
@@ -127,3 +128,10 @@
 	pinctrl-names = "default";
 	status = "ok";
 };
+
+&flexpwm1_pwm0 {
+	status = "okay";
+	pinctrl-0 = <&pinmux_flexpwm1_pwm0A
```
Comment (danieldegrasse) [inline boards/arm/lpcxpresso55s36/lpcxpresso55s36.dts:134]: pin control properties on LPC SOCs are group based, so an array of phandles for `pinctrl-0` properties should not be used here. You should create one `pinctrl-flexpwm1` group, and add the `PWM1_A0` and `PWM1_B0` pins to it

File: drivers/pwm/pwm_mcux.c [hunk 14]
```diff
@@ -135,23 +141,33 @@ static int pwm_mcux_init(const struct device *dev)
...
 		return err;
 	}
-
+#endif
 	PWM_GetDefaultConfig(&pwm_config);
 	pwm_config.prescale = config->prescale;
 	pwm_config.reloadLogic = kPWM_ReloadPwmFullCycle;
 	pwm_config.clockSource = kPWM_BusClock;
 
+	/* Since the Idle function at the system's level of */
```
Comment (danieldegrasse) [inline drivers/pwm/pwm_mcux.c:162]: nit: block comment here should be formatted as:
```
/*
 * Some comment
 */
```
rather that multiple `/* */` styled comments

Also, would prefer some clarification in the comment here. The idle function is putting the SOC itself into a sleep state, and during the power off sequence of the LPC architecture it also puts the PWM into a sleep state, correct?

File: soc/arm/nxp_lpc/lpc55xxx/soc.h (hunk 9 above)
Comment (danieldegrasse) [reply soc/arm/nxp_lpc/lpc55xxx/soc.h:35]: Again, regarding the commit here: Can you make the commit message slightly more verbose? "Added MCUX_PWM dependency; missing funct def; enabled submodule clocks" leaves me wondering what function definition was missing, it is not clear you are talking about a pin function versus a C function until you look at the commit diff.

File: drivers/pwm/pwm_mcux.c [hunk 15]
```diff
@@ -135,23 +141,33 @@ static int pwm_mcux_init(const struct device *dev)
...
-	int i, err;
+	int i;
 
 	if (!device_is_ready(config->clock_dev)) {
 		LOG_ERR("clock control device not ready");
 		return -ENODEV;
 	}
 
-	err = pinctrl_apply_state(config->pincfg, PINCTRL_STATE_DEFAULT);
+#ifdef CONFIG_PINCTRL
```
Comment (danieldegrasse) [inline drivers/pwm/pwm_mcux.c:151]: Regarding the pin control changes: These should either be in a separate commit, or the commit message "drivers: pwm: lpc: Enabled Wait-Enable Bit required for LPC55s36 FlexPWM" should be updated to reflect that you are adding WAITEN bit support as well as making pin control optional for this driver.

File: soc/arm/nxp_lpc/lpc55xxx/soc.c [hunk 16]
```diff
@@ -238,6 +238,17 @@ static int nxp_lpc55xxx_init(const struct device *arg)
 	/* restore interrupt state */
 	irq_unlock(oldLevel);
 
+
+	/* Set the Submodule Clocks for FlexPWM */
+#if CONFIG_BOARD_LPCXPRESSO55S36
```
Comment (danieldegrasse) [inline soc/arm/nxp_lpc/lpc55xxx/soc.c:243]: Make this dependency be on the SOC, not the board, sorry I missed this earlier

File: drivers/pwm/pwm_mcux.c [hunk 17]
```diff
@@ -135,23 +141,35 @@ static int pwm_mcux_init(const struct device *dev)
...
 	PWM_GetDefaultConfig(&pwm_config);
 	pwm_config.prescale = config->prescale;
 	pwm_config.reloadLogic = kPWM_ReloadPwmFullCycle;
 	pwm_config.clockSource = kPWM_BusClock;
 
+	/*
+	 * Since the Idle function of the SOC architecture is
+	 * causing the PWM in the LPC55s36 to Sleep, during a low power cycle,
+	 * without a callback to wake up the PWM again, the team has decided to
+	 * to enable this congif flag by default with the ability
```
Comment (danieldegrasse) [inline drivers/pwm/pwm_mcux.c:166]: minor nit-spelling

File: soc/arm/nxp_lpc/lpc55xxx/soc.c [hunk 18]
```diff
@@ -238,6 +238,17 @@ static int nxp_lpc55xxx_init(const struct device *arg)
 	/* restore interrupt state */
 	irq_unlock(oldLevel);
 
+
+	/* Set the Submodule Clocks for FlexPWM */
+#if CONFIG_SOC_LPC55S36
```
Comment (galak) [inline soc/arm/nxp_lpc/lpc55xxx/soc.c:243]: maybe `#if CONFIG_SOC_LPC55S36 && CONFIG_PWM`

**Review by danieldegrasse — DISMISSED** [review]

File: soc/arm/nxp_lpc/lpc55xxx/soc.c [hunk 19]
```diff
@@ -238,6 +238,17 @@ static int nxp_lpc55xxx_init(const struct device *arg)
 	/* restore interrupt state */
 	irq_unlock(oldLevel);
 
+
+	/* Set the Submodule Clocks for FlexPWM */
+#if CONFIG_SOC_LPC55S36 && CONFIG_PWM
```
Comment (galak) [inline soc/arm/nxp_lpc/lpc55xxx/soc.c:243]: I was being lazy, you probably want this to be:

#if defined(CONFIG_SOC_LPC55S36) && defined(CONFIG_PWM)

File: boards/arm/lpcxpresso55s36/doc/index.rst [hunk 20]
```diff
@@ -66,6 +66,7 @@ The Zephyr configuration for the LPC55S36 board (lpcxpresso55s36) is as follows:
 | CLOCK     | on-chip    | clock_control                       |
 +-----------+------------+-------------------------------------+
 | CAN       | on-chip    | canbus                              |
+| PWM       | on-chip    | pwm                                 |
```
Comment (danieldegrasse) [inline boards/arm/lpcxpresso55s36/doc/index.rst:69]: Will this render correctly in the documentation? I think you need the ----- lines to properly render the table

**Comment by dleach02** [comment]
@galak can you revisit to determine if your blockers have been addressed?

File: soc/arm/nxp_lpc/lpc55xxx/Kconfig.soc [hunk 21]
```diff
@@ -54,6 +54,8 @@ config SOC_LPC55S36
 	select ARM_TRUSTZONE_M
 	select CLOCK_CONTROL
 	select HAS_MCUX_MCAN
+#	select HAS_MCUX_IAP if !TRUSTED_EXECUTION_NONSECURE
```
Comment (mmahadevan108) [inline soc/arm/nxp_lpc/lpc55xxx/Kconfig.soc:57]: This changes looks unrelated to PWM. Please delete.

File: soc/arm/nxp_lpc/lpc55xxx/soc.c [hunk 22]
```diff
@@ -238,6 +238,17 @@ static int nxp_lpc55xxx_init(const struct device *arg)
...
 
+
+	/* Set the Submodule Clocks for FlexPWM */
+#if defined(CONFIG_SOC_LPC55S36) && defined(CONFIG_PWM)
+	SYSCON->PWM0SUBCTL |=
+	  (SYSCON_PWM0SUBCTL_CLK0_EN_MASK | SYSCON_PWM0SUBCTL_CLK1_EN_MASK |
+	   SYSCON_PWM0SUBCTL_CLK2_EN_MASK);
+	SYSCON->PWM1SUBCTL |=
+	  (SYSCON_PWM1SUBCTL_CLK0_EN_MASK | SYSCON_PWM1SUBCTL_CLK1_EN_MASK |
+	   SYSCON_PWM1SUBCTL_CLK2_EN_MASK);
```
Comment (mmahadevan108) [inline soc/arm/nxp_lpc/lpc55xxx/soc.c:249]: Can this portion be added to clock_init function as it looks like clock setup for the PWM module.

File: soc/arm/nxp_lpc/lpc55xxx/soc.c [hunk 23]
```diff
@@ -197,6 +197,17 @@ DT_FOREACH_STATUS_OKAY(nxp_lpc_ctimer, CTIMER_CLOCK_SETUP)
 #endif
 
 #endif /* CONFIG_SOC_LPC55S69_CPU0 */
+
```
Comment (mmahadevan108) [inline soc/arm/nxp_lpc/lpc55xxx/soc.c:200]: Can delete this extra line

File: soc/arm/nxp_lpc/lpc55xxx/soc.c [hunk 24]
```diff
@@ -197,6 +197,17 @@ DT_FOREACH_STATUS_OKAY(nxp_lpc_ctimer, CTIMER_CLOCK_SETUP)
...
+
+	/* Set the Submodule Clocks for FlexPWM */
```
Comment (mmahadevan108) [inline soc/arm/nxp_lpc/lpc55xxx/soc.c:202]: Minor, can delete the leading spaces

**Review by mmahadevan108 — DISMISSED** [review]

File: drivers/pwm/Kconfig.mcux [hunk 25]
```diff
@@ -2,10 +2,21 @@
...
 config PWM_MCUX
 	bool "MCUX PWM driver"
 	default y
 	depends on DT_HAS_NXP_IMX_PWM_ENABLED
 	help
 	  Enable mcux pwm driver.
+
+if PWM_MCUX
+
+config PWM_WAITEN_FLAG
```
Comment (galak) [inline drivers/pwm/Kconfig.mcux:16]: should probably be `PWM_MCUX_WAITEN_FLAG`

Comment (EmilioCBen) [reply drivers/pwm/Kconfig.mcux:16]: Question, do you want me to keep the if endif guards for the PWM_MCUX if I change this config to PWM_MCUX_WAITEN_FLAG?

**Comment by dleach02** [comment]
> Need to figure out where there board specific code should go, but it shouldn't be in the driver.

@galak The board specific code has been removed. I'm going to remove your block as it was pointing to this comment. If you have other concerns please reapply the block and point to the item.

File: boards/arm/lpcxpresso55s36/lpcxpresso55s36-pinctrl.dtsi [hunk 26]
```diff
@@ -42,8 +42,8 @@
 	/* Configures pin routing and optionally pin electrical features. */
 	pinmux_sctimer_default: pinmux_sctimer_default {
 		group0 {
-			pinmux = <SCT0_OUT0_PIO1_4>,
-				<SCT0_OUT2_PIO0_15>;
+		  pinmux = <SCT0_OUT0_PIO1_4>,
+			  <SCT0_OUT2_PIO0_15>;
```
Comment (decsny) [inline boards/arm/lpcxpresso55s36/lpcxpresso55s36-pinctrl.dtsi:46]: nit: use tabs instead of spaces (im actually surprised this passed CI, normally it wants tabs too)

File: boards/arm/lpcxpresso55s36/lpcxpresso55s36-pinctrl.dtsi [hunk 27]
```diff
@@ -57,4 +57,53 @@
...
+		};
+	};
+
+	pinmux_flexpwm0_pwm1: pinmux_flexpwm0_pwm1 {
+		group0 {
+			pinmux = <PWM0_A1_PIO1_6>,
+				<PWM0_B1_PIO1_22>;
+			slew-rate = "standard";
+		};
+	};
```
Comment (decsny) [inline boards/arm/lpcxpresso55s36/lpcxpresso55s36-pinctrl.dtsi:74]: Is there a reason why you don't make a node with 2 subnodes instead of 2 unrelated nodes with this curious naming scheme? I actually don't know, I'm just curious

File: drivers/pwm/Kconfig.mcux [hunk 28]
```diff
@@ -2,10 +2,15 @@
...
 config PWM_MCUX
 	bool "MCUX PWM driver"
 	default y
 	depends on DT_HAS_NXP_IMX_PWM_ENABLED
 	help
 	  Enable mcux pwm driver.
+
+if PWM_MCUX
+
+endif # PWM_MCUX
```
Comment (decsny) [inline drivers/pwm/Kconfig.mcux:16]: why is this here?

File: drivers/pwm/pwm_mcux.c [hunk 29]
```diff
@@ -11,7 +11,11 @@
 #include <zephyr/drivers/clock_control.h>
 #include <soc.h>
 #include <fsl_pwm.h>
+
+#ifdef CONFIG_PINCTRL
 #include <zephyr/drivers/pinctrl.h>
+#endif
```
Comment (decsny) [inline drivers/pwm/pwm_mcux.c:17]: does this really need to be conditionally compiled?

File: boards/arm/lpcxpresso55s36/lpcxpresso55s36-pinctrl.dtsi (hunk 27 above)
Comment (EmilioCBen) [reply boards/arm/lpcxpresso55s36/lpcxpresso55s36-pinctrl.dtsi:74]: These both represent an individual sub-module on the flexpwm that can be configured independently via their independent nodes, while combining these nodes into one node doesn't necessarily lose the ability to configure the sub modules independently, it certainly makes the idea less clear. The naming scheme comes from the dts files which were inspired by other pwm nodes already in Zephyr.

File: drivers/pwm/Kconfig.mcux (hunk 28 above)
Comment (EmilioCBen) [reply drivers/pwm/Kconfig.mcux:16]: Originally not intended, however added when I needed a Kconfig option for the pwm enable-wait bit to be enabled. Due to a recent PR this Kconfig is no longer necessary but the if PWM_MCUX may still be needed for future PRs.

File: boards/arm/lpcxpresso55s36/lpcxpresso55s36-pinctrl.dtsi (hunk 26 above)
Comment (EmilioCBen) [reply boards/arm/lpcxpresso55s36/lpcxpresso55s36-pinctrl.dtsi:46]: I don't know how this slipped by, I can change this right away!

File: drivers/pwm/Kconfig.mcux (hunk 28 above)
Comment (mmahadevan108) [reply drivers/pwm/Kconfig.mcux:16]: I think this can be deleted

**Review by danieldegrasse — APPROVED** [review]

**Review by decsny — APPROVED** [review]

**Review by dleach02 — APPROVED** [review]
